[WIP] Rails 5 compatibility

shaneog opened the pull request:
- [x] Set up testing environment
- [x] Rails 5 (and thus ActiveRecord 5) supports only Ruby 2.2.2+ so we
  exclude lower versions from AR5 testing
- [x] Organise code in `active_record` by version
- [ ] Add Arel 7 and Rails 5 compatibility, whilst maintaining backwards compatibility
- [ ] Deal with deprecation notices cleanly

Fixes #180

shaneog commented:
@edpaget I've opened this PR early to get feedback. I'd be grateful if you could take a quick look.

daniel-rikowski commented:
@shaneog Thank you for working on this. I tried this PR, but without much luck.

What is the reason for overriding `PostgreSQLColumn#type`? In plain Rails this returns `:string` for stringish columns , but now it returns `:'character varying'`. (which breaks Formtastic in my case)

shaneog commented:
Thanks for trying this out @daniel-rikowski!
Initially I used the override because I couldn't find a public API to retrieve column types. However, I think that I've managed to figure out a way around this now, so it should no longer be needed. I'll try to get some more time to work on this today, and push my changes here.

edpaget commented on gemfiles/Gemfile.activerecord-5.0.x:
```diff
@@ -0,0 +1,10 @@
+gem "pg", "~> 0.15"
+
+unless ENV['CI'] || RUBY_PLATFORM =~ /java/
+  gem 'byebug'
```
I haven't personally been using this gem, so if you want to remove it entirely that's fine with me.

edpaget commented on lib/postgres_ext/arel/5.0/visitors/postgresql.rb:
```diff
@@ -0,0 +1,61 @@
+require 'arel/visitors/postgresql'
```
This file looks like a copy of the 4.2 version. Is there are any reason to not just use that one, or are you still working on modifying it for Arel 7?

shaneog replied:
At the moment only `visit_Arel_Nodes_Contains` is different, but I'm still working on modifying it.

shaneog commented on lib/postgres_ext/arel/5.0/predications.rb:
```diff
@@ -0,0 +1,39 @@
+require 'arel/predications'
```
Exactly, I was following the structure previously set, but once I have this working I will refactor as cleanly as possible.

pinak1180 commented:
@shaneog it breaks simple form in my case due to override PostgreSQLColumn#type

KushalP commented:
Hello! Has there been any progress on this? Is there any way I can help?

shaneog commented:
I'll try to get back to it this weekend. I've been a bit busier than expected at work this week.

KushalP commented:
Small point that might useful. Under Rails 5 I get the following deprecation warning running the branch at a9d7b4a:

```
DEPRECATION WARNING: table_exists? is deprecated and will be removed from Rails 5.1 (use #data_source_exists? instead)
(called from call_with_feature at /<snip>/bundler/gems/postgres_ext-a9d7b4abcdbe/lib/postgres_ext/active_record/relation/predicate_builder/array_handler.rb:17)
```

kyoyk commented:
use includes method, report ArgumentError: wrong number of arguments (1 for 2).
I think arel7 compatibility problem.

KushalP commented:
Due to the recent changes to ActiveRecord it looks like the behaviour of using a CTE as the default table has changed slightly. I've written a small commit to make the `where` clause a bit more explicit but would like thoughts: https://github.com/KushalP/postgres_ext/commit/faccda72520ddf9a5c87d0755acfccb4a90813da

It fixes the single test failure that is being produced by the branch at the moment.

shaneog commented:
@daniel-rikowski @pinak1180 This should work for simple_form now since I've reverted the Postgres column type.

@KushalP Thank you for your PR, however I'm going to wait until I've fixed the errors before I remove the deprecation warnings, since I'll refactor the gem more to remove duplicate code.

The only tests that are not passing now are related to `hstore` columns, and the `where_clause_factory` in ActiveRecord 5. When a [hash is passed in](https://github.com/DockYard/postgres_ext/blob/master/test/queries/contains_test.rb#L82) the right side of the relation is a [BindParam](https://github.com/rails/arel/blob/v7.1.1/lib/arel/nodes/bind_param.rb) and not a [Casted](https://github.com/rails/arel/blob/v7.1.1/lib/arel/nodes/casted.rb) Arel Node. 
I'm working on this at the moment, but any help would be appreciated /cc @edpaget

KushalP commented:
> I'm working on this at the moment, but any help would be appreciated

@shaneog Where are you with this? Are there any particular areas you'd like help with? I can't see any work-in-progress commits related to this on your `rails-5` branch.

liamwhite commented:
Bump, waiting to upgrade to Rails 5 until this project is working with it…

subhashb commented:
👍  bump

ltk commented:
In the meantime, https://github.com/gocardless/postgres_ext/tree/rails-5 seems to be working well with Rails 5, at least for our usecases.

liamwhite commented:
@ltk although minor, that fork gives these warnings still:
```
DEPRECATION WARNING: alias_method_chain is deprecated. Please, use Module#prepend instead. From module, you can access the original method using super. (called from <module:QueryMethods> at /home/rails/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bundler/gems/postgres_ext-17e4358fe013/lib/postgres_ext/active_record/relation/query_methods.rb:258)
DEPRECATION WARNING: alias_method_chain is deprecated. Please, use Module#prepend instead. From module, you can access the original method using super. (called from block in <module:ArrayHandlerPatch> at /home/rails/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bundler/gems/postgres_ext-17e4358fe013/lib/postgres_ext/active_record/5.0/relation/predicate_builder/array_handler.rb:28)
```

edpaget commented:
@KushalP do you think we can pull your rails-5 branch in?

KushalP commented:
@edpaget Happy to, are you still one of the maintainers of this gem? I don't want to step on @shaneog's toes with this.

shaneog commented:
@KushalP Don't worry about it! I don't see any functional changes in your branch though - just the deprecation notices are removed in https://github.com/shaneog/postgres_ext/commit/17e4358fe013c05fed530af58cb4bd96225d79fb

@edpaget The issues in [my comment](https://github.com/DockYard/postgres_ext/pull/187#issuecomment-238412113) are still valid. How do you want to deal with them? I don't mind merging @KushalP's changes into this branch and getting it ready for a merge. 

However these branches (both mine and @KushalP's) are missing some significant functionality.

shaneog commented:
Ping @edpaget - I have some time this coming weekend that I can use to put this branch in a mergeable state. Any chance you could take a look at my [questions above](https://github.com/DockYard/postgres_ext/pull/187#issuecomment-238412113)?

edpaget commented:
@shaneog the question about passing a hash in a bind param?

shaneog commented:
That's right @edpaget - it's preventing me making this work for `hstore` columns.

edpaget commented:
Thanks am looking at it

kakoni commented:
GoCardless fork gives me `NoMethodError - undefined method `has_attribute?' for #<CTEProxy:0x007ff2313cf570>` on rails 5.0.0.1.

bosoxbill commented:
Can I help in any way? I'd love to see this get merged.

shaneog commented:
@bosoxbill I got stuck on [these issues](https://github.com/DockYard/postgres_ext/pull/187#issuecomment-238412113) and need some help. I haven't had time to get back to this as my project hasn't yet migrated to Rails 5. 

> The only tests that are not passing now are related to hstore columns, and the where_clause_factory in ActiveRecord 5. When a hash is passed in the right side of the relation is a BindParam and not a Casted Arel Node.

If I can get some help to solve those issues I will clean up and prepare this PR for merging immediately.

jcoleman commented on test/queries/common_table_expression_test.rb:
```diff
@@ -61,7 +61,7 @@
   describe '.from_cte(common_table_expression_hash)' do
     it 'generates an expression with the CTE as the main table' do
-      query = Person.from_cte('lucky_number_seven', Person.where(lucky_number: 7)).where(id: 5)
+      query = Person.from_cte('lucky_number_seven', Person.where(lucky_number: 7)).where(lucky_number_seven: { id: 5 })
```
This is changing the spec to match broken behavior instead of fixing the bug. The `lucky_number_seven` alias is what we should expect to be used by default -- and in fact it is done so in both the SELECT and FROM clauses in the generated query (just not the WHERE) clause.

ms-ati commented:
What's the status on this? Think it'll be finished?

TiSer commented:
Really, guys, how is it going?

shaneog commented:
I've remained stuck on [these issues](https://github.com/DavyJonesLocker/postgres_ext/pull/187#issuecomment-280370707), as previously mentioned.
I simply removed `postgres_ext` from my application during the migration to Rails 5.

@jcoleman thanks for the suggestion - good catch.

I'll close this now, since I'm no longer working on this.

TiSer commented:
So, as for now, "postgrex_ext" is not compatible with Rails 5?

TiSer commented:
Please, get us to know when it will be fixed.

ashasekar20 commented:
I'm facing issue while updating to rails 5.2.0.

Bundler::GemRequireError: There was an error while trying to load the gem 'postgres_ext'.
Gem Load Error is: undefined method `alias_method_chain' for ActiveRecord::QueryMethods:Module
Did you mean?  alias_method
Backtrace for gem load error is:

edpaget commented:
Unless someone else is interested in becoming the maintainer of this Gem is is not no longer being actively maintained and will not be updated further. I will update it's README to that effect.
